### ISSUE TITLE
Generated swagger doc not showing multiple error responses of same error status

### ISSUE THREAD

**Issue by arjunkarnwal** [softwaremill/tapir#497]
I am using Tapir to define my API interface and also to generate swagger open api doc. However I am running into this issue

I have defined my error classes as

```
sealed abstract class SError extends Product with Serializable
object SError {
  final case class Error1(error: String) extends SError
  final case class Error2(error: String) extends SError
  final case class Error3(error: String) extends SError
}
```
I have multiple error out Mapping i.e.

```
EndpointOutput.OneOf[SError] = 
    oneOf(
      statusMapping(StatusCodes.InternalServerError, jsonBody[Error1].description("Internal server error1.")),
      statusMapping(StatusCodes.InternalServerError, jsonBody[Error2].description("Internal server error2.")),
      statusMapping(StatusCodes.NotFound, jsonBody[Error3].description("Not found")),
    )
```
When I generate the swagger doc, in my generated doc I only get 1 InternalServerError description and not two as expected i.e. description for Error2 and not for Error1. I assume it always take last error mapping for same status code.

Is there a way to get the two different description on generated doc ?

**Comment by stsatlantis** [comment]
After checking the code it looks like the problem lies here:

https://github.com/softwaremill/tapir/blob/master/core/src/main/scala/sttp/tapir/internal/package.scala#L96


```scala

private[internal] type BasicOutputsOrMap = Either[BasicOutputs, ListMap[Option[StatusCode], BasicOutputs]]
    private[internal] def asBasicOutputsOrMap: BasicOutputsOrMap = {
      def throwMultipleOneOfMappings = throw new IllegalArgumentException(s"Multiple one-of mappings in output $output")

      def mergeMultiple(v: Vector[BasicOutputsOrMap]): BasicOutputsOrMap = v.foldLeft(Left(Vector.empty): BasicOutputsOrMap) {
        case (Left(os1), Left(os2))    => Left(os1 ++ os2)
        case (Left(os1), Right(osMap)) => Right(osMap.map { case (sc, os2) => sc -> (os1 ++ os2) })
        case (Right(osMap), Left(os2)) => Right(osMap.map { case (sc, os1) => sc -> (os1 ++ os2) })
        case (Right(_), Right(_))      => throwMultipleOneOfMappings
      }

      output match {
        case EndpointOutput.Multiple(outputs)     => mergeMultiple(outputs.map(_.asBasicOutputsOrMap))
        case EndpointIO.Multiple(outputs)         => mergeMultiple(outputs.map(_.asBasicOutputsOrMap))
        case EndpointOutput.Mapped(wrapped, _, _) => wrapped.asBasicOutputsOrMap
        case EndpointIO.Mapped(wrapped, _, _)     => wrapped.asBasicOutputsOrMap
        case _: EndpointOutput.Void               => Left(Vector.empty)
        case s: EndpointOutput.OneOf[_] =>
          Right(
            ListMap(
              s.mappings
                .map(c => (c.output.asBasicOutputsOrMap, c.statusCode))
                .map {
                  case (Left(basicOutputs), statusCode) => statusCode -> basicOutputs
                  case (Right(_), _)                    => throwMultipleOneOfMappings
                }: _*
            )
          )
        case f: EndpointOutput.FixedStatusCode => Right(ListMap(Some(f.statusCode) -> Vector(f)))
        case f: EndpointOutput.StatusCode if f.documentedCodes.nonEmpty =>
          val entries = f.documentedCodes.keys.map(code => Some(code) -> Vector(f)).toSeq
          Right(ListMap(entries: _*))
        case b: EndpointOutput.Basic[_] => Left(Vector(b))
      }
    }

```

The fix can be:

```scala
case s: EndpointOutput.OneOf[_] =>
          Right(
            ListMap(
              (s.mappings
                .map(c => (c.output.asBasicOutputsOrMap, c.statusCode))
                .foldLeft(Map.empty[Option[StatusCode], BasicOutputs]) { (acc, e) =>
                  e match {
                    case (Left(basicOutputs), statusCode) =>
                      acc + (statusCode -> acc.get(statusCode).fold(basicOutputs)(e => e ++ basicOutputs))
                    case (Right(_), _) => throwMultipleOneOfMappings
                  }
                }
                .toList): _*
            )
          )
```

**Comment by arjunkarnwal** [comment]
@stsatlantis Yes I was checking this too last and thats the source of the problem. Thanks

**Comment by stsatlantis** [comment]
@arjunkarnwal Do you agree with this output: https://github.com/softwaremill/tapir/pull/498/files#diff-831740014600e82171c56917d6754fa1
I haven't found any reference in the OpenApi docs which permits the per return type based description in the response. Let me if I'm wrong.

**Comment by arjunkarnwal** [comment]
@stsatlantis Yes looks good. I have left a comment though. More of a question actually.

**Comment by stsatlantis** [comment]
@arjunkarnwal In the meanwhile, there is a workaround:

```scala
sealed abstract class SError extends Product with Serializable
object SError {
  sealed trait InternalServerError extends SError
  final case class Error1(error: String) extends InternalServerError
  final case class Error2(error: String) extends InternalServerError
  final case class Error3(error: String) extends SError
}
```

```scala
EndpointOutput.OneOf[SError] = 
    oneOf(
      statusMapping(StatusCodes.InternalServerError, jsonBody[InternalServerError].description("Internal server error with detailed description."))
      statusMapping(StatusCodes.NotFound, jsonBody[Error3].description("Not found")),
    )
```
Sorry for not getting back sooner. 


I'm working on the other solution, I hope this workaround is feasible for you.

**Comment by markarasev** [comment]
The workaround won't work if you have different types of error bodies for the same status code, e.g. a string message and a detailed JSON. Only one of the different body types will appear in the generated documentation. 
Having different error body types for a same status code may seem strange but it could happen when moving from one type to the other and needing to maintain both versions while clients are migrating.

Any chance https://github.com/softwaremill/tapir/pull/498 will be resumed?

**Comment by adamw** [comment]
What would you expect as the output of the documentation? Is it legal according to the OpenAPI spec to have the same status code a couple of time, when describing the response of a single endpoint?

I think an example yaml which would be the "expected output" would be great.

**Comment by markarasev** [comment]
You're right, this is not legal according to OpenAPI. In fact, the JSON / YAML file formats would not allow this too. 
I think this issue can be closed then. 
Or maybe could we add at least a warning log when someone attemps to convert to OpenAPI an endpoint specification containing multiple status mappings for a same status code instead of silently selecting one over the other?

**Comment by francesconero** [comment]
Sorry, I might have misunderstood the issue, but I think OpenApi 3.0 supports different response schemas with the same status code by using the oneOf syntax.

**Comment by markarasev** [comment]
Yes and this is currently supported by Tapir when you declare a _single_ status mapping for a status code using an error hierarchy like in this [comment](https://github.com/softwaremill/tapir/issues/497#issuecomment-605663391).

But if I understood correctly, the original issue (and mine) were about declaring _several_ status mappings for a single status code and producing several response objects (each of which can use the oneOf syntax) for the same status code, which is actually impossible.

On the other hand, Tapir allows to declare several status mappings for a single status code, but only one will be visible in the Swagger documentation which was kind of surprising for me even if I was familiar with openapi YAML format.

**Comment by francesconero** [comment]
I see, sorry I got confused when you said

_The workaround won't work if you have different types of error bodies for the same status code, e.g. a string message and a detailed JSON._

I understand what you meant now, you can either return different json types or different plain text types but not a mix between different media types, right?

**Comment by markarasev** [comment]
I just checked and I got that wrong too: you can have several status mappings for a single status code if they have different media types and it just works.

The only remaining issue is when you have _several status mappings_ with the _same media type_ for a _single status code_. The status mappings seem to be merged in the generated documentation which can't be the expected behavior.

This code:

```scala
case class Error1(message: String, data: JsObject)
case class Error2(message: String)

val testEndpoint = endpoint.in("test").get.out(stringBody).errorOut(
  oneOf(
    statusMapping(
      StatusCode.BadRequest,
      jsonBody[Error1].description("Bad request - Error 1.")
    ),
    statusMapping(
      StatusCode.BadRequest,
      jsonBody[Error2].description("Bad request - Error 2.")
    )
  )
)

OpenAPIDocsInterpreter.toOpenAPI(testEndpoint, title = "BOWI API", version = "0.1").toYaml
```

produces the following response object in the generated YAML:

```yaml
'400':
  description: Bad request - Error 1.
  content:
    application/json:
      schema:
        $ref: '#/components/schemas/Error2'
```

Here's a reproducer: https://github.com/markarasev/tapir-multiple-errors-same-code 

In my opinion, at the very least a warning log should tell the user she attempts to do something illegal in regard to OpenAPI specification.

The behavior seems to have changed since this issue was originally submitted. I'd suggest to close this issue and to open an new one.

**Comment by adamw** [comment]
To sum up:
* same status code, different media types - currently doesn't work, but is being fixed in #248 (cc @kubinio123 )
* same status code, same media type, here we have a couple of options:
  1. require the user to create a common super-trait and map the response to an inheritance hierarchy. This will result in a `oneOf` schema, usually with a discriminator
  2. accept such responses, and generate a `oneOf`, even though there's no super-trait. However, there won't be a discriminator, so the clients can have problems deciding, which response schema to use
  3. log a warning
  4. throw a runtime exception, if such a mapping is detected

We can run a small poll and decide on the best course of action.

**Comment by adamw** [comment]
As part of #248, we'll go for (iv) - throwing a runtime exception, if a duplicate status code + media type pair is detected, along with supporting multiple media types for the same status code
